cmd/fs-helpers.go

```diff
@@ -16,76 +16,245 @@
-// The staging buffer is then written to the disk. Returns for any error that occurs
-// while reading the stream or writing to disk. Caller should cleanup partial files.
-// Upon errors total data written will be 0 and returns error, on success returns
-// total data written to disk.
-func fsCreateFile(disk StorageAPI, reader io.Reader, buf []byte, bucket, object string) (int64, error) {
-	bytesWritten := int64(0)
-	// Read the buffer till io.EOF and appends data to path at bucket/object.
-	for {
-		n, rErr := reader.Read(buf)
-		if rErr != nil && rErr != io.EOF {
-			return 0, traceError(rErr)
+func fsStatDir(statPath string) (os.FileInfo, error) {
+	fi, err := os.Stat(preparePath(statPath))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errVolumeNotFound
+		} else if os.IsPermission(err) {
+			return nil, errVolumeAccessDenied
 		}
-		bytesWritten += int64(n)
-		wErr := disk.AppendFile(bucket, object, buf[0:n])
-		if wErr != nil {
-			return 0, traceError(wErr)
+		return nil, err
+	}
+	if !fi.IsDir() {
+		return nil, errVolumeAccessDenied
+	}
+	return fi, nil
+}
+
+func fsStatFile(statPath string) (os.FileInfo, error) {
+	fi, err := os.Stat(preparePath(statPath))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, errFileNotFound
+		} else if os.IsPermission(err) {
+			return nil, errFileAccessDenied
+		}
+		return nil, err
+	}
+	return fi, nil
+}
+
+func fsReadFile(readPath string, offset int64) (io.ReadCloser, int64, error) {
+	fr, err := os.Open(preparePath(readPath))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, 0, errFileNotFound
+		} else if os.IsPermission(err) {
+			return nil, 0, errFileAccessDenied
+		} else if isSysErrNotDir(err) {
+			// File path cannot be verified since one of the parents is a file.
+			return nil, 0, errFileAccessDenied
+		} else if isSysErrPathNotFound(err) {
+			// Add specific case for windows.
+			return nil, 0, errFileAccessDenied
+		}
+		return nil, 0, err
+	}
+
+	// Seek to the requested offset.
+	if offset > 0 {
+		_, err = fr.Seek(offset, os.SEEK_SET)
+		if err != nil {
+			return nil, 0, err
 		}
-		if rErr == io.EOF {
-			break
+	}
+
+	// Stat to get the size of the file at path.
+	st, err := fr.Stat()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Success.
+	return fr, st.Size(), nil
+}
+
+// Create file, takes an input reader and staging buffer. Staging buffer
+// is used by io.CopyBuffer to stage the copy operation.
+func fsCreateFile(tempObjPath string, reader io.Reader, buf []byte, fallocSize int64) (int64, error) {
+	writer, err := os.OpenFile(preparePath(tempObjPath), os.O_CREATE|os.O_WRONLY, 0666)
+	if err != nil {
+		// File path cannot be verified since one of the parents is a file.
+		if isSysErrNotDir(err) {
+			return 0, errFileAccessDenied
 		}
+		return 0, err
+	}
+	defer writer.Close()
+	// Fallocate negative, fall back to maximum object size.
+	if fallocSize < 0 {
+		fallocSize = maxObjectSize
```

Comment:
Allocating maxObjectSize when we don't know the real size is aggressive because fallocate *does* consume space in the disk in case of Linux and we'll end up allocating much more space than needed.